### PR TITLE
Boltdb

### PR DESCRIPTION
**Pull request by sharadregoti**
Testing is remaianing

### CODE REVIEW

**Comment by YourTechBud** [comment]
Test the library first. Let's get this off the ground. Also, your tests seem to be failing

File: gateway/go.mod [hunk 1]
```diff
@@ -4,6 +4,7 @@ require (
 	cloud.google.com/go v0.38.0
 	github.com/armon/go-metrics v0.0.0-20190430140413-ec5e00d3c878 // indirect
 	github.com/aws/aws-sdk-go v1.19.7
+	github.com/boltdb/bolt v1.3.1
```
Comment (YourTechBud) [inline gateway/go.mod]: You are using the wrong library. Use `bbolt` https://github.com/etcd-io/bbolt

File: gateway/modules/crud/bolt/aggregate.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package bolt
+
+import (
+	"context"
+
+	"github.com/spaceuptech/space-cloud/gateway/model"
+)
+
+func (b *Bolt) Aggregate(ctx context.Context, project, col string, req *model.AggregateRequest) (interface{}, error) {
+	return nil, nil
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/aggregate.go]: Send an error that it isn't supported

File: gateway/modules/crud/bolt/batch.go [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package bolt
+
+import (
+	"context"
+
+	"github.com/spaceuptech/space-cloud/gateway/model"
+)
+
+func (b *Bolt) Batch(ctx context.Context, project string, req *model.BatchRequest) ([]int64, error) {
+	return nil, nil
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/batch.go]: Send error

File: gateway/modules/crud/bolt/bolt.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+)
+
+type Bolt struct {
+	enabled    bool
+	connection string
+	client     *bolt.DB
+}
+
+// Init initialises a new bolt instance
+func Init(enabled bool, connection string) (mongoStub *Bolt, err error) {
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/bolt.go]: why is the variable here called mongo stub?

File: gateway/modules/crud/bolt/bolt.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+	return nil
+}
+
+func (b *Bolt) connect() error {
+	client, err := bolt.Open(b.connection, 0600, bolt.DefaultOptions)
+	if err != nil {
+		return err
+	}
+
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/bolt.go]: log success message here. Do this for all dbs as well

File: gateway/modules/crud/bolt/collections.go [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package bolt
+
+import (
+	"context"
+
+	"github.com/spaceuptech/space-cloud/gateway/utils"
+)
+
+func (b *Bolt) GetCollections(ctx context.Context, project string) ([]utils.DatabaseCollections, error) {
+	return nil, nil
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/collections.go]: return error

File: gateway/modules/crud/bolt/collections.go [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+func (b *Bolt) DeleteCollection(ctx context.Context, project, col string) error {
+	return nil
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/collections.go]: return error

File: gateway/modules/crud/bolt/create.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+func (b *Bolt) Create(ctx context.Context, project, col string, req *model.CreateRequest) (int64, error) {
+	// Create a collection object
+	// todo check if exists
+	switch req.Operation {
+	case utils.One:
+		// Insert single document
+		if err := b.client.Update(func(tx *bolt.Tx) error {
+			count, _, err := b.Read(ctx, project, col, &model.ReadRequest{
+				Find:      req.Document.(map[string]interface{}),
+				Operation: utils.All,
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/create.go]: Use the operation count here. Will save you from all the unmarshalling overload

File: gateway/modules/crud/bolt/create.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+// Create inserts a document (or multiple when op is "all") into the database
+func (b *Bolt) Create(ctx context.Context, project, col string, req *model.CreateRequest) (int64, error) {
+	// Create a collection object
+	// todo check if exists
+	switch req.Operation {
+	case utils.One:
+		// Insert single document
+		if err := b.client.Update(func(tx *bolt.Tx) error {
+			count, _, err := b.Read(ctx, project, col, &model.ReadRequest{
+				Find:      req.Document.(map[string]interface{}),
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/create.go]: Just check if the `_id` exists. Don't check for the entire document, that is incorrect. What if the user insert two documents with the same `_id` but different `name` fields

File: gateway/modules/crud/bolt/create.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+// Create inserts a document (or multiple when op is "all") into the database
+func (b *Bolt) Create(ctx context.Context, project, col string, req *model.CreateRequest) (int64, error) {
+	// Create a collection object
+	// todo check if exists
+	switch req.Operation {
+	case utils.One:
+		// Insert single document
+		if err := b.client.Update(func(tx *bolt.Tx) error {
+			count, _, err := b.Read(ctx, project, col, &model.ReadRequest{
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/create.go]: If you are using the `count` operation. The `count` variable will always be one, even if no documents were present. You having to probe the result in this case.

File: gateway/modules/crud/bolt/create.go [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+				logrus.Error("error inserting into bboltdb data already exists - %v", err)
+			}
+
+			b, err := tx.CreateBucketIfNotExists([]byte(project))
+			if err != nil {
+				logrus.Errorf("error creating bucket in bboltdb while inserting- %v", err)
+				return err
+			}
+			id, ok := req.Document.(map[string]interface{})["_id"]
+			if ok {
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/create.go]: Check for the `!ok` condition and throw error there only

File: gateway/modules/crud/bolt/create.go [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+				value, err := json.Marshal(req.Document)
+				if err != nil {
+					logrus.Errorf("error marshalling while inserting in bboltdb - %v", err)
+					return err
+				}
+
+				if err = b.Put([]byte(fmt.Sprintf("%s/%s", col, id)), value); err != nil {
+					return err
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/create.go]: Why no log error here

File: gateway/modules/crud/bolt/create.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+				}
+			}
+			logrus.Errorf("error bboltdb unable to find _id while inserting - %v", err)
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/create.go]: According to your logic this will always be executed. Not good. Use the `!ok` system i suggested above

File: gateway/modules/crud/bolt/create.go [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+	"github.com/boltdb/bolt"
+	"github.com/sirupsen/logrus"
+
+	"github.com/spaceuptech/space-cloud/gateway/model"
+	"github.com/spaceuptech/space-cloud/gateway/utils"
+)
+
+// Create inserts a document (or multiple when op is "all") into the database
+func (b *Bolt) Create(ctx context.Context, project, col string, req *model.CreateRequest) (int64, error) {
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/create.go]: You have seperated out the implementation of `all` and `one` operations. I would suggest you only implement the `all` operation. Convert the object to an array if it was a `one` operation. Let's keep things DRY.

File: gateway/modules/crud/bolt/create.go [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+				}
+
+				if err = b.Put([]byte(fmt.Sprintf("%s/%s", col, id)), value); err != nil {
+					return err
+				}
+			}
+			logrus.Errorf("error bboltdb unable to find _id while inserting - %v", err)
+			return err
+		}); err != nil {
+			return 0, nil
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/create.go]:
```suggestion
			return 0, err
```

File: gateway/modules/crud/bolt/create.go [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+					return err
+				}
+
+				if err = bucket.Put([]byte(fmt.Sprintf("%s/%s", col, id)), value); err != nil {
+					return err
+				}
+			}
+			return nil
+		}); err != nil {
+			return 0, nil
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/create.go]:
```suggestion
			return 0, err
```

File: gateway/modules/crud/bolt/raw.go [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+// GetConnectionState : function to check connection state
+func (b *Bolt) GetConnectionState(ctx context.Context) bool {
+	if !b.enabled || b.client == nil {
+		return false
+	}
+
+	// Ping to check if connection is established
+	err := b.connect()
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/raw.go]: Is there a better way to do this?

File: gateway/modules/crud/bolt/describe.go [hunk 18]
```diff
@@ -0,0 +1,11 @@
+package bolt
+
+import (
+	"context"
+
+	"github.com/spaceuptech/space-cloud/gateway/utils"
+)
+
+func (b *Bolt) DescribeTable(ctc context.Context, project, col string) ([]utils.FieldType, []utils.ForeignKeysType, []utils.IndexType, error) {
+	return nil, nil, nil, nil
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/describe.go]: return error

File: gateway/modules/crud/bolt/delete.go [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+				}
+				// if valid then update
+				if utils.Validate(req.Find, result) {
+					// delete data
+					if err := bucket.Delete(k); err != nil {
+						return err
+					}
+					// exit the loop
+					break
+				}
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/delete.go]: Just have one function for both operations. Break if op was one else continue

File: gateway/modules/crud/bolt/delete.go [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+
+			// get all keys matching the prefix
+			prefix := []byte(col + "/")
+			for k, v := c.Seek(prefix); k != nil && bytes.HasPrefix(k, prefix); k, v = c.Next() {
+				result := map[string]interface{}{}
+				if err := json.Unmarshal(v, result); err != nil {
+					logrus.Errorf("error un marshalling while reading from bboltdb - %v", err)
+					return err
+				}
+				// if valid then update
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/delete.go]:
```suggestion
				// if valid then delete
```

File: gateway/modules/crud/bolt/raw.go [hunk 21]
```diff
@@ -0,0 +1,26 @@
+package bolt
+
+import "context"
+
+func (b *Bolt) RawExec(ctx context.Context, project string) error {
+	return nil
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/raw.go]: return error in all unsupported operations

File: gateway/modules/crud/bolt/read.go [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+	"github.com/boltdb/bolt"
+	"github.com/sirupsen/logrus"
+
+	"github.com/spaceuptech/space-cloud/gateway/model"
+	"github.com/spaceuptech/space-cloud/gateway/utils"
+)
+
+func (b *Bolt) Read(ctx context.Context, project, col string, req *model.ReadRequest) (int64, interface{}, error) {
+
+	switch req.Operation {
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/read.go]: We can have a single implementation for one an all

File: gateway/modules/crud/bolt/read.go [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+
+		var count int64
+		results := []interface{}{}
+		if err := b.client.View(func(tx *bolt.Tx) error {
+			// Assume bucket exists and has keys
+			c := tx.Bucket([]byte(project)).Cursor()
+
+			prefix := []byte(col + "/")
+			for k, v := c.Seek(prefix); k != nil && bytes.HasPrefix(k, prefix); k, v = c.Next() {
+				count++
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/read.go]: Don't update the count here. Update the count where you amend the the result

File: gateway/modules/crud/bolt/read.go [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+					logrus.Errorf("error un marshalling while reading from bboltdb - %v", err)
+					return err
+				}
+				if utils.Validate(req.Find, result) {
+					results = append(results, result)
+				}
+			}
+			return nil
+		}); err != nil {
+			return 0, nil, nil
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/read.go]:
```suggestion
			return 0, nil, err
```

File: gateway/modules/crud/bolt/delete.go [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+					count++
+					// delete data
+					if err := bucket.Delete(k); err != nil {
+						return err
+					}
+				}
+			}
+			return nil
+		}); err != nil {
+			return 0, nil
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/delete.go]:
```suggestion
			return 0, err
```

File: gateway/modules/crud/bolt/read.go [hunk 26]
```diff
@@ -0,0 +1,70 @@
...
+					logrus.Errorf("error un marshalling while reading from bboltdb - %v", err)
+					return err
+				}
+				if utils.Validate(req.Find, result) {
+					finalResult = result
+				}
+			}
+			return nil
+		}); err != nil {
+			return 0, nil, nil
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/read.go]:
```suggestion
			return 0, nil, err
```

File: gateway/modules/crud/bolt/update.go [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+	"github.com/sirupsen/logrus"
+
+	"github.com/spaceuptech/space-cloud/gateway/model"
+	"github.com/spaceuptech/space-cloud/gateway/utils"
+)
+
+func (b *Bolt) Update(ctx context.Context, project, col string, req *model.UpdateRequest) (int64, error) {
+
+	switch req.Operation {
+	case utils.One, utils.Upsert:
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/update.go]: Upsert acts as regular update `all` operation when there one or more documents already exist which match the find clause

File: gateway/modules/crud/bolt/update.go [hunk 28]
```diff
@@ -0,0 +1,95 @@
...
+			prefix := []byte(col + "/")
+			for k, v := c.Seek(prefix); k != nil && bytes.HasPrefix(k, prefix); k, v = c.Next() {
+				result := map[string]interface{}{}
+				if err := json.Unmarshal(v, result); err != nil {
+					logrus.Errorf("error un marshalling while reading from bboltdb - %v", err)
+					return err
+				}
+				// if valid then update
+				if utils.Validate(req.Find, result) {
+					value, err := json.Marshal(req.Update)
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/update.go]: req,Update contains a map like the following:
```json
{
  "$set": {"name": "John"}
}
```
You need to selectively update fields in the current object, and overwrite using the updated object

File: gateway/modules/crud/bolt/update.go [hunk 29]
```diff
@@ -0,0 +1,95 @@
...
+					if err = bucket.Put(k, value); err != nil {
+						return err
+					}
+					// exit the loop
+					break
+				}
+			}
+			return nil
+		}); err != nil {
+			return 0, nil
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/update.go]:
```suggestion
			return 0, err
```

File: gateway/modules/crud/bolt/update.go [hunk 30]
```diff
@@ -0,0 +1,95 @@
...
+
+					// over ride the data
+					if err = bucket.Put(k, value); err != nil {
+						return err
+					}
+				}
+			}
+			return nil
+		}); err != nil {
+			return 0, nil
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/update.go]:
```suggestion
			return 0, err
```

File: gateway/modules/crud/bolt/raw.go (hunk 17 above)
Comment (sharadregoti) [reply gateway/modules/crud/bolt/raw.go]: image
we do have some utility function which can be used, if succesfully executed we can assume it is client is still connected with the database

File: gateway/modules/crud/bolt/create.go (hunk 10 above)
Comment (sharadregoti) [reply gateway/modules/crud/bolt/create.go]: i don't think so,
my implementatin 
image

Comment (YourTechBud) [reply gateway/modules/crud/bolt/create.go]: You've implement count incorrectly. Check the implementation for other databases

File: gateway/modules/crud/bolt/update.go [hunk 31]
```diff
@@ -0,0 +1,89 @@
...
+						// exit the loop
+						break
+					}
+				}
+			}
+			return nil
+		}); err != nil {
+			return 0, err
+		}
+		// todo $set should be map or array ?
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/update.go]: remove this comment

File: gateway/modules/crud/bolt/update.go [hunk 32]
```diff
@@ -0,0 +1,89 @@
...
+		objToSet, ok := req.Update["$set"].(map[string]interface{})
+		if !ok {
+			return 0, fmt.Errorf("error unable to update in bbolt - $set db operator not found or the operator value is not map")
+		}
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/update.go]: put this inside the if

File: gateway/modules/userman/operations.go [hunk 33]
```diff
@@ -27,7 +27,7 @@ func (m *Module) Profile(ctx context.Context, token, dbType, project, id string)
 		return 0, nil, err
 	}
 	switch utils.DBType(actualDbType) {
-	case utils.Mongo:
+	case utils.Mongo, utils.BoltDb:
```
Comment (YourTechBud) [inline gateway/modules/userman/operations.go]: It should be `utils.BoltDB`

File: gateway/modules/crud/bolt/raw.go [hunk 34]
```diff
@@ -0,0 +1,27 @@
...
+	"context"
+	"fmt"
+)
+
+func (b *Bolt) RawExec(ctx context.Context, project string) error {
+	return fmt.Errorf("error raw exec cannot be performed over selected database")
+}
+
+func (b *Bolt) CreateProjectIfNotExist(ctx context.Context, project string) error {
+	return fmt.Errorf("errocr create project operation cannot be performed over selected database")
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/raw.go]: Error

File: gateway/modules/crud/bolt/read.go [hunk 35]
```diff
@@ -0,0 +1,73 @@
...
+					if req.Operation == utils.One {
+						break
+					}
+				}
+			}
+			return nil
+		}); err != nil {
+			return 0, nil, err
+		}
+		if req.Operation == utils.One && count == 1 {
```
Comment (YourTechBud) [inline gateway/modules/crud/bolt/read.go]: Return error if operation was one but count is zero

File: gateway/utils/handlers/filestore.go [hunk 36]
```diff
@@ -124,6 +124,16 @@ func HandleRead(auth *auth.Module, fileStore *filestore.Module) http.HandlerFunc
 			}
 			json.NewEncoder(w).Encode(map[string]interface{}{"result": res})
 			return
+		} else if op == "exist" {
+			log.Println("ldsfilh", path)
```
Comment (YourTechBud) [inline gateway/utils/handlers/filestore.go]: Why is this here

File: gateway/utils/handlers/filestore.go [hunk 37]
```diff
@@ -124,6 +124,16 @@ func HandleRead(auth *auth.Module, fileStore *filestore.Module) http.HandlerFunc
...
+			if err := fileStore.DoesExists(ctx, project, token, path); err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
```
Comment (YourTechBud) [inline gateway/utils/handlers/filestore.go]: Return not found status code
